### DISCUSSION TITLE
Can't use asyncio.sleep in gather?

### DISCUSSION THREAD

**Discussion by skyguy** [micropython/micropython#13043] [Using MicroPython] [answered]
I tried to add a sleep() to a gather() but MicroPython v1.21 gives me an error "coroutine expected"? Am I doing it wrong?

Simplified example:
```py
import asyncio

async def main():
    await asyncio.gather(asyncio.sleep(5))
    print("done")

asyncio.run(main())
```
```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "gather_sleep_tst.py", line 7, in <module>
  File "asyncio/core.py", line 1, in run
  File "asyncio/core.py", line 1, in run_until_complete
  File "asyncio/core.py", line 1, in run_until_complete
  File "gather_sleep_tst.py", line 4, in main
  File "asyncio/funcs.py", line 1, in gather
  File "asyncio/funcs.py", line 1, in <listcomp>
  File "asyncio/core.py", line 1, in _promote_to_task
  File "asyncio/core.py", line 1, in create_task
TypeError: coroutine expected
```

Thanks in advance
-Kevin

**Comment by mendenm** [comment]
shouldn't that  be a 1-item tuple (or a list):
```python
import asyncio

async def main():
    await asyncio.gather((asyncio.sleep(5),))
    print("done")

asyncio.run(main())
```
since gather wants an iterable?

**Reply by peterhinch** [reply]
To avoid confusion `gather` does not want an iterable. It takes an arbitrary number of positional args each of which is an `awaitable`. If `aws` is a sequence of `awaitable` objects the call signature is
```py
awaitable asyncio.gather(*aws, return_exceptions=False)  # aws is unpacked
```

**Reply by skyguy** [reply]
Thanks for the clarification, and a special thanks for all the tutorials and clarifications you've published. Most of my questions while learning MicroPython were answered by something you wrote!

**Comment by jimmo** [comment, marked as answer]
@skyguy For efficiency, asyncio.sleep is a bit "magic" and isn't implemented as a true coroutine. You'll need to wrap it in a task to make this work.

(It's worth it though, this optimisation means you can do `await asyncio.sleep` without causing an allocation)

**Reply by skyguy** [reply]
OK, thanks. I'll try it that way.

**Comment by mendenm** [comment]
Oops, my apology.  I should have re-checked.

Marcus
